File: python/packages/botas-fastapi/src/botas_fastapi/bot_app.py
"""Zero-boilerplate bot host using FastAPI + Uvicorn.

Example::

    from botas_fastapi import BotApp

    app = BotApp()

    @app.on("message")
    async def on_message(ctx):
        await ctx.send(f"You said: {ctx.activity.text}")

    app.start()
"""

# NOTE: Do NOT use `from __future__ import annotations` here — it breaks
# FastAPI's runtime inspection of endpoint parameter type hints (Request).

import os
from contextlib import asynccontextmanager
from typing import Any, Awaitable, Callable

from botas.bot_application import BotApplication, InvokeResponse
from botas.core_activity import CoreActivity, ResourceResponse
from botas.i_turn_middleware import TurnMiddleware
from botas.token_manager import BotApplicationOptions
from botas.turn_context import TurnContext

from botas_fastapi.bot_auth import bot_auth_dependency

ActivityHandler = Callable[[TurnContext], Awaitable[None]]


class BotApp:
    """Composes a :class:`BotApplication` with a FastAPI server.

    Provides ``on()``, ``use()``, and ``start()`` so a bot can be created
    and running in just a few lines of code.
    """

    def __init__(
        self,
        options: BotApplicationOptions = BotApplicationOptions(),
        *,
        port: int | None = None,
        path: str = "/api/messages",
        auth: bool | None = None,
    ) -> None:
        self.bot = BotApplication(options)
        self._port = port
        self._path = path
        self._auth = auth

    def on(
        self,
        type: str,
        handler: "ActivityHandler | None" = None,
    ) -> Any:
        """Register a handler for an activity type.

        Works as a decorator or a two-argument call — delegates to
        :meth:`BotApplication.on`.
        """
        return self.bot.on(type, handler)

    def on_invoke(
        self,
        name: str,
        handler: "Callable[[TurnContext], Awaitable[InvokeResponse]] | None" = None,
    ) -> Any:
        """Register a handler for an invoke activity by its ``activity.name`` sub-type.

        Works as a decorator or a two-argument call — delegates to
        :meth:`BotApplication.on_invoke`.
        """
        return self.bot.on_invoke(name, handler)

    def use(self, middleware: TurnMiddleware) -> "BotApp":
        """Register a middleware in the turn pipeline."""
        self.bot.use(middleware)
        return self

    async def send_activity_async(
        self,
        service_url: str,
        conversation_id: str,
        activity: "CoreActivity | dict[str, Any]",
    ) -> "ResourceResponse | None":
        """Proactively send an activity to a conversation."""
        return await self.bot.send_activity_async(service_url, conversation_id, activity)

    def _build_app(self) -> Any:
        """Build and return the FastAPI application (without starting it)."""
        from fastapi import Depends, FastAPI, Request

        @asynccontextmanager
        async def lifespan(app):
            # Startup
            yield
            # Shutdown - close the bot's HTTP client
            # CORS is not needed: Bot Service calls this endpoint directly (no browser)
            await self.bot.aclose()

        auth_enabled = self._auth if self._auth is not None else bool(self.bot.appid)
        deps = [Depends(bot_auth_dependency(self.bot.appid))] if auth_enabled else []

        fastapi_app = FastAPI(lifespan=lifespan)
        bot = self.bot
        path = self._path

        @fastapi_app.post(path, dependencies=deps)
        async def messages(request: Request) -> Any:
            from fastapi import HTTPException
            from fastapi.responses import JSONResponse

            body = await request.body()
            if len(body) > 10 * 1024 * 1024:  # 10 MB limit
                raise HTTPException(status_code=413, detail="Request body too large")
            try:
                invoke_response = await bot.process_body(body.decode())
            except ValueError as exc:
                raise HTTPException(status_code=400, detail=str(exc)) from exc
            if invoke_response is not None:
                content = invoke_response.body if invoke_response.body is not None else {}
                return JSONResponse(content=content, status_code=invoke_response.status)
            return {}

        @fastapi_app.get("/")
        async def root() -> dict:
            app_id = bot.appid or "(no client ID)"
            return {"message": f"Bot {app_id} Running - send messages to {path}"}

        @fastapi_app.get("/health")
        async def health() -> dict:
            return {"status": "ok"}

        return fastapi_app

    def start(self) -> None:
        """Build the FastAPI app and start Uvicorn. Blocks until shutdown."""
        import uvicorn

        app = self._build_app()
        port = self._port or int(os.environ.get("PORT", "3978"))
        uvicorn.run(app, host="0.0.0.0", port=port)

File: node/packages/botas-express/src/bot-app.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import type { Server } from 'node:http'
import express from 'express'
import {
  BotApplication,
  type BotApplicationOptions,
  type CoreActivity,
  type CoreActivityHandler,
  type InvokeActivityHandler,
  type TurnMiddleware,
  type ResourceResponse,
  getLogger,
} from 'botas-core'
import { botAuthExpress } from './bot-auth-express.js'

/**
 * Options for {@link BotApp}.
 *
 * Extends the core {@link BotApplicationOptions} with Express-specific
 * settings for port, route path, and auth toggle.
 */
export interface BotAppOptions extends BotApplicationOptions {
  /** Port to listen on. Defaults to the `PORT` env var or `3978`. */
  port?: number
  /** Path for the messages endpoint. Defaults to `"/api/messages"`. */
  path?: string
  /**
   * Whether to enable inbound JWT auth.
   * Defaults to `true` when `CLIENT_ID` is configured, `false` otherwise.
   */
  auth?: boolean
}

/**
 * Zero-boilerplate bot host that composes a {@link BotApplication} with an
 * Express server.
 *
 * @example
 * ```ts
 * import { BotApp } from 'botas-express'
 *
 * const app = new BotApp()
 * app.on('message', async (ctx) => {
 *   await ctx.send('you said: ' + ctx.activity.text)
 * })
 * app.start()
 * ```
 */
export class BotApp {
  /** The underlying BotApplication instance. */
  readonly bot: BotApplication

  private readonly _options: BotAppOptions

  constructor (options: BotAppOptions = {}) {
    this._options = options
    this.bot = new BotApplication(options)
  }

  /**
   * Register a handler for an activity type.
   * Delegates to {@link BotApplication.on}.
   */
  on (type: string, handler: CoreActivityHandler): this {
    this.bot.on(type, handler)
    return this
  }

  /**
   * Register a handler for an invoke activity by its `activity.name` sub-type.
   * Delegates to {@link BotApplication.onInvoke}.
   */
  onInvoke (name: string, handler: InvokeActivityHandler): this {
    this.bot.onInvoke(name, handler)
    return this
  }

  /**
   * Register a middleware in the turn pipeline.
   * Delegates to {@link BotApplication.use}.
   */
  use (middleware: TurnMiddleware): this {
    this.bot.use(middleware)
    return this
  }

  /**
   * Proactively send an activity to a conversation.
   * Delegates to {@link BotApplication.sendActivityAsync}.
   */
  async sendActivityAsync (
    serviceUrl: string,
    conversationId: string,
    activity: Partial<CoreActivity>
  ): Promise<ResourceResponse | undefined> {
    return this.bot.sendActivityAsync(serviceUrl, conversationId, activity)
  }

  /**
   * Start the Express server.
   *
   * Creates routes for the bot messages endpoint, a health check, and a
   * root status page, then begins listening.
   *
   * @returns The underlying `http.Server` — call `.close()` to shut down.
   */
  start (): Server {
    const port = this._options.port ?? Number(process.env['PORT'] ?? 3978)
    const path = this._options.path ?? '/api/messages'
    const authEnabled = this._options.auth ?? !!this.bot.options.clientId

    const app = express()

    if (authEnabled) {
      app.post(path, botAuthExpress(this.bot.options.clientId), (req, res) => {
        this.bot.processAsync(req, res)
      })
    } else {
      app.post(path, (req, res) => {
        this.bot.processAsync(req, res)
      })
    }

    app.all(path, (_req, res) => {
      res.status(405).set('Allow', 'POST').json({ error: 'Method Not Allowed' })
    })

    app.get('/health', (_req, res) => { res.json({ status: 'ok' }) })
    app.get('/', (_req, res) => {
      res.send(`Bot ${this.bot.options.clientId ?? '(no client ID)'} is running. Send messages to ${path}`)
    })

    const server = app.listen(port, () => {
      getLogger().info(`BotApp listening on http://localhost:${port}${path}`)
    })

    return server
  }
}
